src/sage/knots/free_knotinfo_monoid.py

```diff
@@ -0,0 +1,479 @@
+#       Copyright (C) 2024 Sebastian Oehms <seb.oehms@gmail.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 2 of the License, or
+# (at your option) any later version.
+#                  http://www.gnu.org/licenses/
+##############################################################################
+
+from sage.knots.knotinfo import SymmetryMutant
+from sage.monoids.indexed_free_monoid import IndexedFreeAbelianMonoid, IndexedFreeAbelianMonoidElement
+from sage.misc.cachefunc import cached_method
+
+
+class FreeKnotInfoMonoidElement(IndexedFreeAbelianMonoidElement):
+    """
+    An element of an indexed free abelian monoid.
+    """
+    def as_knot(self):
+        r"""
+        Return the knot represented by ``self``.
+
+        EXAMPLES::
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: FKIM.inject_variables(select=3)
+            Defining K3_1
+            Defining K3_1m
+            sage: K = K3_1^2 * K3_1m
+            sage: K.as_knot()
+            Knot represented by 9 crossings
+        """
+        wl = self.to_word_list()
+        P = self.parent()
+        if len(wl) == 1:
+            name = wl[0]
+            L = P._index_dict[name][0].link()
+            if name.endswith(SymmetryMutant.mirror_image.value):
+                return L.mirror_image()
+            if name.endswith(SymmetryMutant.reverse.value):
+                return L.reverse()
+            if name.endswith(SymmetryMutant.concordance_inverse.value):
+                return L.mirror_image().reverse()
+            return L
+        else:
+            from sage.misc.misc_c import prod
+            return prod(P.gen(wl[i]).as_knot() for i in range(len(wl)))
+
+    def to_knotinfo(self):
+        r"""
+        Return a word representing ``self`` as a list of pairs where each pair
+        ``(ki, sym)`` consists of a :class:`~sage.knots.knotinfo.KontInfoBase` instance ``ki`` and
+        :class:`~sage.knots.knotinfo.SymmetryMutant` instance ``sym``.
+
+        EXAMPLES::
+
+            sage: from sage.knots.free_knotinfo_monoid import FreeKnotInfoMonoid
+            sage: FKIM =  FreeKnotInfoMonoid()
+            sage: FKIM.inject_variables(select=3)
+            Defining K3_1
+            Defining K3_1m
+            sage: K = K3_1^2 * K3_1m
+            sage: K.to_knotinfo()
+            [(<KnotInfo.K3_1: '3_1'>, <SymmetryMutant.itself: 's'>),
+            (<KnotInfo.K3_1: '3_1'>, <SymmetryMutant.itself: 's'>),
+            (<KnotInfo.K3_1: '3_1'>, <SymmetryMutant.mirror_image: 'm'>)]
+        """
+        wl = self.to_word_list()
+        P = self.parent()
+        return [P._index_dict[w] for w in wl]
+
+
+class FreeKnotInfoMonoid(IndexedFreeAbelianMonoid):
```

Comment:
Perhaps this should also inherit from `Singleton`?

Reply:
This is fixed by the current commit.